### PR TITLE
feat(sync): optimize the cascading deletion of managed member cluster resources

### PR DESCRIPTION
**Pull request by qclc**
This PR is meant to replace https://github.com/kubewharf/kubeadmiral/pull/225

### CODE REVIEW

File: pkg/controllers/sync/controller.go [hunk 1]
```diff
@@ -1104,58 +1110,31 @@ func (s *SyncController) reconcileClusterForCascadingDeletion(
...
-			} else if err != nil && !meta.IsNoMatchError(err) && !apierrors.IsNotFound(err) {
-				remainingByGVK[gvk] = fmt.Sprintf("Unknown (failed to list from cluster: %v)", err)
-			}
-		}
+	// concurrent cascading-delete operations for each ftc
+	wg := sync.WaitGroup{}
+	cascadingDeletionStatus := sync.Map{}
+	for i := range ftcs {
+		wg.Add(1)
+		go s.cascadingDeletionForFTC(ctx, ftcs[i], &cascadingDeletionStatus, &wg, cluster.Name)
```
Comment (gary-lgy) [inline pkg/controllers/sync/controller.go:1131]: `wg` doesn't have to be an argument to `cascadingDeletionForFTC`

File: pkg/controllers/sync/controller.go [hunk 2]
```diff
@@ -1186,3 +1155,140 @@ func (s *SyncController) reconcileClusterForCascadingDeletion(
...
+		client, err := s.getClusterClient(clusterName)
+		if err != nil {
+			return nil, err
+		}
+
+		objects, err := client.Resource(ftc.GetSourceTypeGVR()).Namespace(corev1.NamespaceAll).List(ctx,
+			metav1.ListOptions{
+				LabelSelector: labels.SelectorFromSet(labels.Set{
+					managedlabel.ManagedByKubeAdmiralLabelKey: managedlabel.ManagedByKubeAdmiralLabelValue,
+				}).String(),
```
Comment (gary-lgy) [inline pkg/controllers/sync/controller.go:1261]: Should we list with ResourceVersion: "0"?

File: pkg/controllers/sync/controller.go [hunk 3]
```diff
@@ -1186,3 +1155,140 @@ func (s *SyncController) reconcileClusterForCascadingDeletion(
...
+	if len(clusterObjs) == 0 {
+		return
+	}
+
+	// cascading-delete managed cluster objects
+	var deleteFailedObjs []string
+	for i := range clusterObjs {
+		clusterObj := clusterObjs[i].DeepCopy()
+		namespacedName := fmt.Sprintf("%s/%s", clusterObj.GetNamespace(), clusterObj.GetName())
+		_, keyedLogger := logging.InjectLoggerValues(ctx, "namespacedName", namespacedName)
```
Comment (gary-lgy) [inline pkg/controllers/sync/controller.go:1186]: Let's inject these separately

File: pkg/controllers/sync/controller.go [hunk 4]
```diff
@@ -1186,3 +1155,140 @@ func (s *SyncController) reconcileClusterForCascadingDeletion(
...
+			continue
+		}
+
+		// execute the cascading-delete operation
+		dispatcher := dispatch.NewUnmanagedDispatcher(s.getClusterClient, fedObj.TargetGVR(), fedObj.TargetName(), s.metrics)
+		s.removeFromCluster(ctx, dispatcher, clusterName, fedObj, clusterObj, true)
+
+		dispatchOk, timeoutErr := dispatcher.Wait()
+		if timeoutErr != nil {
+			keyedLogger.Error(timeoutErr, "Cascading-delete cluster object timeout")
```
Comment (gary-lgy) [inline pkg/controllers/sync/controller.go:1212]: Do we need to append to `deleteFailedObjs`?

File: pkg/controllers/sync/controller.go (hunk 1 above)
Comment (qclc) [reply pkg/controllers/sync/controller.go]: The code has been updated.

File: pkg/controllers/sync/controller.go (hunk 2 above)
Comment (qclc) [reply pkg/controllers/sync/controller.go:1264]: I think we should. The code has been updated.

File: pkg/controllers/sync/controller.go (hunk 3 above)
Comment (qclc) [reply pkg/controllers/sync/controller.go]: The code has been updated.

File: pkg/controllers/sync/controller.go (hunk 4 above)
Comment (qclc) [reply pkg/controllers/sync/controller.go:1216]: It is not needed here. If `timeoutErr != nil`, `dispatchOk` must be false, so `deleteFailedObjs` is added directly under the judgment statement of `dispatchOk`. 
Additional comments are added here to explain the relationship between timeoutErr and dispatchOk.

File: pkg/controllers/sync/controller.go [hunk 5]
```diff
@@ -1186,3 +1155,140 @@ func (s *SyncController) reconcileClusterForCascadingDeletion(
...
+		}
+		if !dispatchOk {
+			keyedLogger.Error(nil, "Failed to cascading-delete target object")
+			deleteFailedObjs = append(deleteFailedObjs, namespacedName)
+		}
+	}
+
+	// if the update operation fails, do not continue the delete operation,
```
[the remark this answers is not in the document]
Comment (qclc) [reply pkg/controllers/sync/controller.go]: Yes, the code has been updated.

**Review by gary-lgy — APPROVED** [review]

File: pkg/controllers/sync/controller.go [hunk 6]
```diff
@@ -581,6 +580,16 @@ func (s *SyncController) syncToClusters(
 			}
 		}
 
+		// If cascading deletion is triggered,
+		// wait for reconcileClusterForCascadingDeletion to perform the deletion operation.
+		// If the delete operation has not been performed, federatedObject's status will be updated to WaitingForRemoval.
```
Comment (gary-lgy) [inline pkg/controllers/sync/controller.go:585]:
```suggestion
		// If the delete operation has not been performed, federatedObject's status will be updated to WaitingForCascadingDeletion.
```

Comment (qclc) [reply pkg/controllers/sync/controller.go]: The code has been updated.

**Review by limhawjia — APPROVED** [review]

**Review by mrlihanbo — APPROVED** [review]

**Pull request merged**